### PR TITLE
Reduce LOH allocs when applying code fixes.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Drops LOH allocs from 1.1GB:

image

to around 50mb:

image

### CODE REVIEW

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/AbstractLinkedFileMergeConflictCommentAdditionService.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@ internal abstract class AbstractLinkedFileMergeConflictCommentAdditionService :
 {
     internal abstract string GetConflictCommentText(string header, string beforeString, string afterString);
 
-    public IEnumerable<TextChange> CreateEdits(SourceText originalSourceText, IEnumerable<UnmergedDocumentChanges> unmergedChanges)
+    public List<TextChange> CreateEdits(SourceText originalSourceText, List<UnmergedDocumentChanges> unmergedChanges)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/AbstractLinkedFileMergeConflictCommentAdditionService.cs:19]: changed a lot of the APIs to not be IEnumerable-bases (as it's impossible to tell what is lazy and waht isn't).  this also means things like enumerations don't allocate.

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs [hunk 2]
```diff
@@ -125,12 +126,13 @@ internal async Task<LinkedFileMergeSessionResult> MergeDiffsAsync(IMergeConflict
...
         var unmergedDocumentChanges = new List<TextChange>();
-        var successfullyMergedChanges = ArrayBuilder<TextChange>.GetInstance();
+        using var _ = ArrayBuilder<TextChange>.GetInstance(out var successfullyMergedChanges);
 
         var cumulativeChangeIndex = 0;
 
-        var textchanges = await textDiffService.GetTextChangesAsync(oldDocument, newDocument, cancellationToken).ConfigureAwait(false);
-        foreach (var change in textchanges)
+        var textChanges = await textDiffService.GetTextChangesAsync(
+            oldDocument, newDocument, TextDifferenceTypes.Line, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:134]: switching to Line-based-diff is a major win here.  We normally do word-based, which leads to *massive* allocations in hte VS side (they're the impl of this service).

**Comment by CyrusNajmabadi** [comment]
@ToddGrun  ptal.

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/UnmergedDocumentChanges.cs [hunk 3]
```diff
@@ -9,9 +9,9 @@
 
 namespace Microsoft.CodeAnalysis;
 
-internal sealed class UnmergedDocumentChanges(IEnumerable<TextChange> unmergedChanges, string projectName, DocumentId documentId)
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/UnmergedDocumentChanges.cs:12]: >internal [](http://example.com/codeflow?start=0&length=8)

might as well clear out the #nullable in this file too

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs [hunk 4]
```diff
@@ -196,25 +198,25 @@ internal async Task<LinkedFileMergeSessionResult> MergeDiffsAsync(IMergeConflict
...
         {
             unmergedChanges.Add(new UnmergedDocumentChanges(
-                unmergedDocumentChanges.AsEnumerable(),
+                unmergedDocumentChanges,
                 oldDocument.Project.Name,
                 oldDocument.Id));
         }
 
-        return successfullyMergedChanges.ToImmutableAndFree();
+        return successfullyMergedChanges.ToImmutable();
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:209]: >ToImmutable [](http://example.com/codeflow?start=41&length=11)

Haven't you been primarily using ToImmutableAndClear in these situations?

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:209]:
```suggestion
        return successfullyMergedChanges.ToImmutableAndClear();
```

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs [hunk 5]
```diff
@@ -309,7 +309,7 @@ public void LogLinkedFileResult(LinkedFileGroupSessionInfo info)
             => LinkedFileGroups.Add(info);
     }
 
-    internal class LinkedFileGroupSessionInfo
+    internal sealed class LinkedFileGroupSessionInfo
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingSession.cs:312]: >sealed [](http://example.com/codeflow?start=13&length=6)

seal the class right above this too?

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**